llmfoundry/data/packing.py

```diff
@@ -260,124 +261,95 @@ def pad_tensor(tensor: torch.Tensor, pad_value: int):
+        dist.all_reduce(packing_ratio_tensor, reduce_operation='MIN')
+        packing_ratio = packing_ratio_tensor.item()
+    return packing_ratio
+
+
+def profile_packing(
+        dataloader_cfg: DictConfig, tokenizer: PreTrainedTokenizerBase,
+        min_ratio: float, max_ratio: float, num_packing_ratios: int,
+        device_batch_size: int) -> Iterable[Tuple[float, float, float]]:
+    """Generator function that profiles example packing across packing ratios.
+
+    Args:
+        dataloader_cfg (DictConfig): The dataloader configuration for profiling.
+        tokenizer (PreTrainedTokenizerBase): The tokenizer for profiling.
+        min_ratio (float): Smallest packing_ratio to test. Must be >=1.
+        max_ratio (float): Largest packing_ratio to test. Must be larger than `min_ratio`.
+        num_packing_ratios (int): Number of packing_ratio values (spaced between `min_ratio` and `max_ratio`) to try.
+        device_batch_size (int): The size of the batches (number of examples) per device.
+
+    Returns:
+        An iterable of tuples of packing ratio, padding, and waste.
+    """
+    import copy
+
+    from llmfoundry.data.dataloader import build_dataloader
+
+    max_seq_len = dataloader_cfg.dataset.get('max_seq_len')
+    max_leftovers_to_keep = dataloader_cfg.dataset.get('max_leftovers_to_keep',
+                                                       None)
 
-    with open(args.yaml_path) as f:
-        cfg = om.load(f)
-    if 'parameters' in cfg:
-        cfg = om.to_container(cfg.parameters)
-        cfg = om.create(cfg)
-    device_batch_size = cfg.global_train_batch_size // args.num_devices
+    # Turn off packing for the dataloader (we want raw, pre-packed examples)
+    dataloader_cfg = copy.deepcopy(dataloader_cfg)
+    dataloader_cfg.dataset.packing_ratio = None
+    dataloader_cfg.dataset.max_leftovers_to_keep = None
+    dataloader_cfg.drop_last = False
+    dataloader_cfg.num_workers = 0
+    dataloader_cfg.prefetch_factor = None
 
     # Determine the packing_ratio values we'll try
     packing_ratios, raw_batch_sizes = [], []
-    for packing_ratio in np.linspace(args.min,
-                                     args.max,
-                                     args.num_packing_ratios,
+    for packing_ratio in np.linspace(min_ratio,
+                                     max_ratio,
+                                     num_packing_ratios,
                                      endpoint=True):
         packing_ratio = np.round(10 * packing_ratio) / 10
         raw_batch_size = int(packing_ratio * device_batch_size)
         if raw_batch_size not in raw_batch_sizes:
             packing_ratios.append(packing_ratio)
             raw_batch_sizes.append(raw_batch_size)
 
-    # Fetch a bunch of raw examples once, which we'll re-use
-    if 'train_loader' not in cfg:
-        raise ValueError('config must define train_loader')
-    dataloader_cfg = cfg.train_loader
+    n_profile_examples = max(raw_batch_sizes) * 100
```

Comment:
This controls how many samples we are going to use for profiling, yeah? So this says we will profile `device_batch_size * max_ratio * 100 samples`. I think you may have already run this, but could you comment here for long this takes to run, and how changing this to 200, 300, 400 changes how long it takes to run? Would help to make an informed decision about this number.

Reply:
I think we should keep it at 100 because it scales linearly and with 100, we're looking at about several minutes to profile. (Based on profiling on dolly)

Reply:
If anything I think we could optimize by using the dataloader with more workers etc and not get a single large batch but mutiple smaller batches. I can create a jira to investigate optimizations here

Reply:
hmm ok. sounds fine for now. please make the jira for further optimization and put it in foundry backlog epic